File: src/ShowyEdge/objc/Updater.m
#ifdef USE_SPARKLE
@import Sparkle;
#endif

#import "Updater.h"

@interface Updater ()

#ifdef USE_SPARKLE
@property SUUpdater* suupdater;
#endif

@end

@implementation Updater

- (instancetype)init {
  self = [super init];

  if (self) {
#ifdef USE_SPARKLE
    self.suupdater = [SUUpdater new];
#endif
  }

  return self;
}

- (NSString*)getFeedURL:(BOOL)includingBetaVersions {
  // ----------------------------------------
  // check beta & stable releases.

  // Once we check appcast.xml, SUFeedURL is stored in a user's preference file.
  // So that Sparkle gives priority to a preference over Info.plist,
  // we overwrite SUFeedURL here.
  if (includingBetaVersions) {
    return @"https://appcast.pqrs.org/showyedge-appcast-devel.xml";
  }

  return @"https://appcast.pqrs.org/showyedge-appcast.xml";
}

- (void)check:(BOOL)isBackground {
#ifdef USE_SPARKLE
  NSString* url = [self getFeedURL:NO];
  [self.suupdater setFeedURL:[NSURL URLWithString:url]];

  NSLog(@"checkForUpdates %@", url);
  if (isBackground) {
    [self.suupdater checkForUpdatesInBackground];
  } else {
    [self.suupdater checkForUpdates:nil];
  }
#endif
}

- (void)checkForUpdatesInBackground {
  [self check:YES];
}

- (void)checkForUpdatesStableOnly {
#ifdef USE_SPARKLE
  NSString* url = [self getFeedURL:NO];
  [self.suupdater setFeedURL:[NSURL URLWithString:url]];
  NSLog(@"checkForUpdates %@", url);
  [self.suupdater checkForUpdates:nil];
#endif
}

- (void)checkForUpdatesWithBetaVersion {
#ifdef USE_SPARKLE
  NSString* url = [self getFeedURL:YES];
  [self.suupdater setFeedURL:[NSURL URLWithString:url]];
  NSLog(@"checkForUpdates %@", url);
  [self.suupdater checkForUpdates:nil];
#endif
}

@end

File: src/ShowyEdge/swift/Updater.swift
import Foundation

#if USE_SPARKLE
  import Sparkle
#endif

final class Updater: ObservableObject {
  static let shared = Updater()

  #if USE_SPARKLE
    private let updaterController: SPUStandardUpdaterController
    private let delegate = SparkleDelegate()
  #endif

  @Published var canCheckForUpdates = false

  init() {
    #if USE_SPARKLE
      updaterController = SPUStandardUpdaterController(
        startingUpdater: true,
        updaterDelegate: delegate,
        userDriverDelegate: nil
      )

      updaterController.updater.publisher(for: \.canCheckForUpdates)
        .assign(to: &$canCheckForUpdates)
    #endif
  }

  func checkForUpdatesInBackground() {
    #if USE_SPARKLE
      delegate.includingBetaVersions = false
      updaterController.updater.checkForUpdatesInBackground()
    #endif
  }

  func checkForUpdatesStableOnly() {
    #if USE_SPARKLE
      delegate.includingBetaVersions = false
      updaterController.checkForUpdates(nil)
    #endif
  }

  func checkForUpdatesWithBetaVersion() {
    #if USE_SPARKLE
      delegate.includingBetaVersions = true
      updaterController.checkForUpdates(nil)
    #endif
  }

  #if USE_SPARKLE
    private class SparkleDelegate: NSObject, SPUUpdaterDelegate,
      SPUStandardUserDriverDelegate
    {
      var includingBetaVersions = false

      func feedURLString(for updater: SPUUpdater) -> String? {
        var url = "https://appcast.pqrs.org/showyedge-appcast.xml"
        if includingBetaVersions {
          url = "https://appcast.pqrs.org/showyedge-appcast-devel.xml"
        }

        print("feedURLString \(url)")

        return url
      }
    }
  #endif
}
